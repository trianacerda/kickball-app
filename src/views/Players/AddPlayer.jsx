import { useState, useEffect } from 'react';
import { useHistory, Link } from 'react-router-dom';
import PlayerForm from '../../components/PlayerForm';
import { createPlayer } from '../../services/players';
import { getTeams } from '../../services/teams';

function AddPlayer() {
  const [name, setName] = useState('');
  const [position, setPosition] = useState('');
  const [selectedTeam, setSelectedTeam] = useState('');
  const [teams, setTeams] = useState([]);
  const history = useHistory();

  useEffect(() => {
    async function getAllTeams() {
      const fetchTeams = await getTeams();
      setTeams(fetchTeams);
    }
    getAllTeams();
  });

  useEffect(() => {
    async function getSelectedTeam() {
      if (!selectedTeam) return;

      if (selectedTeam !== 'Please Select a Team') {
        setSelectedTeam(selectedTeam);
      }
    }
    getSelectedTeam();
  }, [selectedTeam]);

  const onSubmit = async (e) => {
    e.preventDefault();
    const teamId = selectedTeam;
    const response = await createPlayer({ name, position, teamId });
    history.push(`/players/${response[0].id}`);
  };

  return (
    <>
      <h1>Add a Playa!</h1>
      <Link
        to="/teams"
        className="teams-link"
        exact
        style={{ display: 'flex', flexDirection: 'start' }}
      >
        Teams
      </Link>
      <Link
        to="/players"
        className="player-link"
        exact
        style={{ display: 'flex', flexDirection: 'start' }}
      >
        Players
      </Link>
      <PlayerForm
        name={name}
        position={position}
        submitHandler={onSubmit}
        setName={setName}
        setPosition={setPosition}
        teamChange={setSelectedTeam}
        selectedTeam={selectedTeam}
        teams={teams}
      />
    </>
  );
}

export default AddPlayer;
