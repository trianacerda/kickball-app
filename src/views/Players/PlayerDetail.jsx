import { useEffect, useState } from 'react';
import { NavLink } from 'react-router-dom';
import { getPlayerById } from '../../services/players';

function PlayerDetail({ match }) {
  const { playerId } = match.params;
  const [player, setPlayer] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    getPlayerById(playerId)
      .then((resp) => setPlayer(resp))
      .finally(() => setLoading(false));
  }, [playerId]);

  if (loading) return <h1>Please wait, your selected Player is loading...</h1>;

  return (
    <>
      <NavLink to="/" className="home-link" exact>
        Home
      </NavLink>
      <NavLink to="/teams" className="team-link" exact>
        Teams
      </NavLink>
      <NavLink to="/players" className="player-link" exact>
        Players
      </NavLink>
      <h1>{player.name}</h1>
      <ul>
        <li>Position: {player.position}</li>
        <li>Team: {player.teams.name}</li>
        <li>
          From: {player.city}, {player.state}
        </li>
      </ul>
    </>
  );
}

export default PlayerDetail;
