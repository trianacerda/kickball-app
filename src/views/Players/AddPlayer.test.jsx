import { render, screen } from '@testing-library/react';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import { MemoryRouter } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import userEvent from '@testing-library/user-event';
import AddPlayer from './AddPlayer';
import PlayerDetail from './PlayerDetail';

const mockPlayer = {
  created_at: '2021-12-12T21:16:29.0838+00:00',
  id: 29,
  name: 'T-Money',
  position: 'Big Kicker',
  team_id: 3,
  teams: {
    created_at: '2021-12-08T19:24:13+00:00',
    id: 3,
    name: "Luna's Looney Tunes",
    city: 'Happy',
    state: 'OR',
  },
};

const mockTeam = {
  created_at: '2021-12-08T19:24:13+00:00',
  id: 3,
  name: "Luna's Looney Tunes",
  city: 'Happy',
  state: 'OR',
  players: [],
};

const server = setupServer(
  rest.get('https://eyxqfyzkrrtcxhvkkviw.supabase.co/rest/v1/teams', (req, res, ctx) => {
    return res(ctx.json([mockTeam]));
  }),
  rest.post('https://eyxqfyzkrrtcxhvkkviw.supabase.co/rest/v1/teams', (req, res, ctx) => {
    return res(ctx.json([mockTeam]));
  }),
  rest.get('https://eyxqfyzkrrtcxhvkkviw.supabase.co/rest/v1/players', (req, res, ctx) => {
    return res(ctx.json(mockPlayer));
  }),
  rest.post('https://eyxqfyzkrrtcxhvkkviw.supabase.co/rest/v1/players', (req, res, ctx) => {
    return res(ctx.json([mockPlayer]));
  })
);

beforeAll(() => {
  server.listen();
});

afterAll(() => {
  server.close();
});

it('should render Form component and redirect to details part after submit', async () => {
  const history = createMemoryHistory();
  history.push('/players/new');

  const { container } = render(
    <MemoryRouter>
      <AddPlayer />
      <PlayerDetail match={{ params: { playerId: 29 } }} />
    </MemoryRouter>
  );

  screen.getByText('Add a Playa!');

  const nameField = screen.getByLabelText(/Name:/i);
  const positionField = screen.getByLabelText(/Position:/i);
  const submitButton = screen.getByRole('button', { name: 'Submit' });

  userEvent.type(nameField, 'T-Money');
  userEvent.type(positionField, 'Big Kicker');
  userEvent.selectOptions(
    await screen.findByRole('combobox'),
    await screen.findByRole('option', [3])
  );
  userEvent.click(submitButton);

  expect(container).toMatchSnapshot();
  await screen.findByText(/T-Money/i);
});
