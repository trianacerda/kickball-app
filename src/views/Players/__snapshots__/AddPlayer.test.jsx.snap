// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Form component and redirect to details part after submit 1`] = `
<div>
  <h1>
    Add a Playa!
  </h1>
  <a
    class="teams-link"
    href="/teams"
    style="display: flex; flex-direction: start;"
  >
    Teams
  </a>
  <a
    class="player-link"
    href="/players"
    style="display: flex; flex-direction: start;"
  >
    Players
  </a>
  <fieldset>
    <form>
      <label
        for="name"
      >
        Name:
      </label>
      <input
        id="name"
        name="name"
        type="text"
        value="T-Money"
      />
      <label
        for="position"
      >
        Position:
      </label>
      <input
        id="position"
        name="position"
        position="position"
        type="text"
        value="Big Kicker"
      />
      <select>
        <option
          value="select-team"
        >
          Please Select a Team
        </option>
        <option
          value="3"
        >
          Luna's Looney Tunes
        </option>
      </select>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </fieldset>
  <a
    aria-current="page"
    class="home-link active"
    href="/"
    style="display: flex; flex-direction: start;"
  >
    Home
  </a>
  <a
    class="team-link"
    href="/teams"
    style="display: flex; flex-direction: start;"
  >
    Teams
  </a>
  <a
    class="player-link"
    href="/players"
    style="display: flex; flex-direction: start;"
  >
    Players
  </a>
  <h1>
    T-Money
  </h1>
  <ul>
    <li>
      Position: 
      Big Kicker
    </li>
    <li>
      Team: 
      Luna's Looney Tunes
    </li>
    <li>
      From: 
      Happy
      , 
      OR
    </li>
  </ul>
</div>
`;
